omegaconf/omegaconf.py

```diff
@@ -372,21 +372,68 @@ def _unescape_word_boundary(match: Match[str]) -> str:
         return [re.sub(r"(\\([ ,]))", lambda x: x.group(2), x) for x in escaped]
 
     @staticmethod
-    def register_resolver(name: str, resolver: Resolver) -> None:
+    def register_resolver(
+        name: str,
+        resolver: Resolver,
+        args_as_strings: bool = True,
+        use_cache: Optional[bool] = True,
+    ) -> None:
+        """
+        The `args_as_strings` flag was introduced to preserve backward compatibility
+        with the older resolver system, which assumed that resolvers took the raw string
+        representation of their inputs:
+            - `True` is the old behavior (the resolver uses the string representation
+              of its inputs), and triggers a warning
+            - `False` is the new behavior (the resolver can take non-string inputs), and
+              will become the default in the future
+        """
```

Comment:
1. args_as_strings: 
this is referring only to args specified as literals in the interpolation, right?
e.g  `a` and `10` here `${foo:a,10}`?

If so, the user is in full control over the input and we would eventually be able to remove this.

if an arg is an interpolation, is the value being converted to a string too?
${foo:${bar}}

If we want to remove args_as_strings, what is the deprecation plan?

Reply:
Currently this is applying to all arguments, so when using `${foo:${bar}}` with resolver `foo` registered with `args_as_strings=True`, then `foo` would receive `"${bar}"` as input. As far as I'm concerned this is ok because currently using `{foo:${bar}}` has a very weird behavior that people shouldn't be using anyway -- it returns the string `"${foo:xxx}"` where `xxx` is the resolution of `${bar}`. So it shouldn't break normal code, and if someone tries to do this in this new version they will get the warning about how they should set `args_as_strings=False`.

My proposition for deprecation plan is the following:
1. In 2.2, change default value of `args_as_strings` to `False` and add a deprecation warning that this argument will be removed in a future version (so users should just remove it in their code)
2. In 2.3, make `args_as_strings=True` crash
3. In 2.4, remove `args_as_strings`

Step 2 could be skipped if 2.3 is released a very long time after 2.2.

Reply:
Conversation mostly continued in https://github.com/omry/omegaconf/pull/321#discussion_r513628009

One point this makes me realize though is that with 2.1 supporting `legacy_register_resolver()`, people may expect nested interpolations to work with it. I'll add a basic check to raise an exception if an input string containing `${` is found in the arguments of a resolver registered with `legacy_register_resolver()`, to force people to switch to `new_register_resolver()` if they want to nest interpolations.

Reply:
> I'll add a basic check to raise an exception if an input string containing `${` is found in the arguments of a resolver registered with `legacy_register_resolver()`, to force people to switch to `new_register_resolver()` if they want to nest interpolations.

Done in 45665c1